### ISSUE TITLE
OpenMined Grid ®

### ISSUE THREAD

**Issue by bendecoste** [OpenMined/OpenMined#299]
### Motivation

We have been talking about having the ability to train a bunch of different models a bunch of different ways quickly.  Enter the OpenMined Grid!

Imagine being able to train a model testing multiple different architectures, learning rates, etc all at the same time, well soon you can!

### The Plan

Subject to change based on discussion in here!

![image](https://user-images.githubusercontent.com/1411145/34579470-8685f226-f15f-11e7-8e95-1ef4b720afbb.png)

Clients will make an API call which contains the all the configurations that they want to run.  The API is definitely up for debate and not finalized so please comment with your ideas.

```python
var highDropout = Sequential([
    nn.Linear(3,4),
    nn.ReLU(),
    nn.Dropout(0.10),
    nn.Linear(4,2),
    nn.Softmax(1),
    nn.Log()
])

var lowDropout = Sequential([
    nn.Linear(3,4),
    nn.ReLU(),
    nn.Dropout(0.05),
    nn.Linear(4,2),
    nn.Softmax(1),
    nn.Log()
])

grid.learn(
  input=input # see [1]
  target=target # see [1]
  models=[highDropout, lowDropout]
  learningRates=[1, 0.1, 0.01, 0.001]
)
```

[1] - the input/target is likely going to be an IPFS address that contains contains the learning data.  In my opinion, it makes sense for the user API to take tensor objects like normal, and it will serialize & write to IPFS behind the scenes.  This makes it easier for us to switch from IPFS down the road if needed, and hides an unnecessary thing to learn from the user.

Once the smart contract is complete it will write back the results to IPFS and the client gets notified.

**Comment by bendecoste** [comment]
Some other thoughts.  The above api doesn't really let you mix and match configs that well.  In the example above I would expect that to learn 8 times.  Each learning rate with each model.  But maybe you don't want to test fast learning rates with the high dropout model.  

Another possible api

```python
grid.learn(
  input = input
  target = target
  configurations = [config1, config2, config3]
)
```

where the configration objects look something like this

```python
config1 = grid.configuration(
  model=highDropout,
  learningRate=0.01
)

config2 = gird.configuration(
  model=lowDropout,
  learningRate=1
)

config3 = grid.configuration(
  model=lowDropout,
  learningRate=0.1
)
```

**Comment by bendecoste** [comment]
### Division of labour

I am pretty interested in owning the IPFS portion of this project.  I think that the other big pieces are

- blockchain
- implementing the API itself (both client and server)

**Comment by justin1121** [comment]
I can look into the python API. I like your second API suggestion. Will probably do something like that!

**Comment by bendecoste** [comment]
cool .. i believe @gavinuhma is looking into blockchain piece

**Comment by jeamick** [comment]
Hi guys , fantastic if someone needs help for infura or anything concerning Blockchain feel free to ping me on slack @justin1121  @bendecoste

**Comment by bendecoste** [comment]
This is done!
